### ISSUE TITLE
New web interface

### ISSUE THREAD

**Issue by tananaev** [traccar/traccar-web#592]
Apparently Sencha is dying:

https://www.sencha.com/forum/showthread.php?377323-Sencha-sold-to-IDERA

I think we need to start seriously thinking of rewriting the web app.

**Comment by tananaev** [comment]
One potential library that we can use:

https://dojotoolkit.org/community/roadmap/

**Comment by Abyss777** [comment]
Bad news, I was waiting for ExtJS 7.
Any way, It will take much time to implement the same level of functionality. I think we have to develop it in parallel for some time.

**Comment by tananaev** [comment]
Agree. There is still some chances that they release version 7, but I think we should start working on alternative.

**Comment by Abyss777** [comment]
Sure, I'll start to studying this framework.
It would be nice If you create some skeleton, may be branch or subfolder

**Comment by tananaev** [comment]
We need to decide what to use first. Maybe we should use react with something like this:

http://www.material-ui.com/

**Comment by tananaev** [comment]
Good thing about react is that we can potentially make native mobile apps later using react native.

**Comment by Abyss777** [comment]
We are used to working with stores on ExtJS and dojo also has similar thing https://dojotoolkit.org/reference-guide/1.7/dojo/store.html#dojo-store
I think it is very important benefit.

What if we create list of features waited from framework... like
- [ ] Openlayers support
- [ ] RESTful data stores
- [ ] Tables with filters
- [ ] ...

**Comment by tananaev** [comment]
OpenLayers is just a map library. You can include it in a simple div.

REST data stores and tables with filters are nice to have, but not really a requirement. We can implement all that ourselves. The benefit would be that we have more control and it will probably be much faster than ExtJS stuff.

I think the main requirement is to have enough UI widgets for everything that we need. Grid, forms with all different field types, various toolbars and menus, resizable panels. Last one needs more research. Possibly we can just use some pure HTML components.

**Comment by tananaev** [comment]
Another option for nice UI widgets:

https://getmdl.io/

It's not really a JS framework, but it's fast and look nice.

Demo that might be kind of relevant to our use case:

https://getmdl.io/templates/dashboard/index.html

**Comment by tananaev** [comment]
I have also found this. Might be useful as a reference on how others implement similar functionality:

https://github.com/geoadmin/mf-geoadmin3

**Comment by tananaev** [comment]
Another thing I want to consider is using Leaflet library for maps. It's much smaller (god for mobile) and simpler than OpenLayers. There is a good plugin for including Google Maps (a lot of people are interested in this). It doesn't have as many features as OpenLayers, but then we don't actually need that many features. We use markers and various geozones and that's about it. What do you think?

**Comment by Abyss777** [comment]
Implementing stores, tables and filters might take a time... But agree ExtJS is not fast at all.

I briefly looked at Leaflet, I like it at first sight. It has all we need and what is interesting I found a couple of plugins for SVG marker icons, I believe we could implement native SVG icons, and manipulate them as SVG, without rerendering to <img> on all change.

Also I liked mf-geoadmin3, very nice looking. Good reference.

One important thing to decide is licence. Will we migrate back to Apache 2.0? Are these libraries fit by licence to us?

**Comment by tananaev** [comment]
Yes, that's one of the most important parts. We want Apache 2.0 again. All of the libraries that I've mentioned should be good.

I want to try React with Material UI to see if it would work for us.

**Comment by Abyss777** [comment]
OK. React is popular and used/supported by big companies, do not think it will die soon. Material UI is nice looking. 
Hope you finally will like tabs :smiley:

**Comment by Abyss777** [comment]
One more, I'm not good in web programming, but as I understand React is only "V" in MVC, what will we use for states? Redux?

**Comment by tananaev** [comment]
I'm not an expert myself, so I guess we'll have to learn.

**Comment by Abyss777** [comment]
One more thing also would be nice to have is unit tests.

**Comment by tananaev** [comment]
Agree, it would be nice to have some tests for web app.

**Comment by ovnicraft** [comment]
IMO project can move: react + material-ui + Leaflet; best options its has compatible licenses.
Said that, you will start a plan for ?

**Comment by tananaev** [comment]
I don't get the question.

**Comment by tananaev** [comment]
I have done some more investigation and it seems like React is not a very good match for our project. Neither Leaflet nor OpenLayers have a good integration with it. There are only some very basic wrappers. I'm sure it's possible to somehow make React and map libraries work together, but it's not going to be easy or nice.

**Comment by Abyss777** [comment]
Agree, I wrappers (I have found) are very basic. I believe they do not support all we need.

**Comment by tananaev** [comment]
What do you think about following stack:

- [Material Components](https://material.io/components/web/) for UI
- [Leaflet](http://leafletjs.com/) library for map
- Pure modern JavaScript for most other stuff
- Some smaller third party libraries as required

**Comment by Abyss777** [comment]
I like it, a lot of work, but we could do it smaller and faster without frameworks.
Only thing, I could not find tables (grids in terms of ExtJS) in Material Components.

**Comment by tananaev** [comment]
I have noticed it as well. Apparently it's in development. Maybe we can use some third party component until they develop official one.

https://github.com/material-components/material-components-web/issues/57

**Comment by Abyss777** [comment]
Agree, also we may rethink interface and implement most things without tables. They are necessary only for reports.

**Comment by tananaev** [comment]
I was also thinking about it. We can use lists, especially if we want to make it more mobile-friendly.

**Comment by ovnicraft** [comment]
Did you check vueJS ? 
I am not JS expert btw let something here: https://github.com/PaulLeCam/react-leaflet

**Comment by tananaev** [comment]
@ovnicraft, we can consider vue.js if you can explain the benefits.

I have already researched react-leaflet. It is very basic integration as I mentioned above.

Please don't just throw in some names that you heard unless you know what you are talking about.

**Comment by allenhwkim** [comment]
I am outsider

I just saw your conversation linked from Material Design Component data table issue and makes me interested into the discussion. I think the choice you are making is very considerate and admirable. I want to leave some of my experience if you don't mind.

* Material Components for UI

I have been researching for 3 days about good web components. 
Material Components is my best choice compared to bootstrap, Material UI(React), MDL, or Materialize. It's designed to work as component itself regardless of any web framework.

* Leaflet library for map

I have been using Google map for long time and it has very good API, even with streetview etc. I am also the owner of https://ngmap.github.io/ and https://github.com/ng2-ui/map. However you need to pay for it when it becomes commercial. Also it does not provide a very customized requirements, e.g. setting North to any side. 
I used Openlayer. It's powerful, but not developer-friendly IMO.
Not experienced with Leaflet though. :(

* Pure modern JavaScript for most other stuff

ES6/ES7 ways makes the code a lot to easy to read IMO. Also using classes makes documentation a lot easier.

* Some smaller third party libraries as required

You may need a framework, which handles routing, storage, http call handling, server-side-rendering, performance. Experience with angular1, angular2, and react was ok, but not satisfactory. Recently I learned VueJs, and it seems very easy and promising.

**Comment by tananaev** [comment]
Thanks for your comments.

Google Maps are not free, so it's not really an option.

**Comment by Abyss777** [comment]
https://habrahabr.ru/post/334676/
Might be helpful comparing of Leaflet and Openlayers (in Russian)

**Comment by tananaev** [comment]
Doesn't seem like there is anything against Leaflet in the article. There are some negative comments about plugins, but I guess we'll use minimum of plugins and mostly standard API.

**Comment by Abyss777** [comment]
As I understand most features are in plugins, what is in openlayers core is in leaflet plugins, but I believe we will cope with it.

Could not find some timely library to work with REST and local storage.
https://github.com/lincolnloop/amygdala fits but outdated

Other candidate is Dojo again https://dojotoolkit.org/reference-guide/1.10/dojo/store/JsonRest.html

Any other ideas? Or may be again will consider dojo?

**Comment by tananaev** [comment]
We can just use pure JavaScript or jQuery.

**Comment by Abyss777** [comment]
Rewrote `simple` on leaflet

![image](https://user-images.githubusercontent.com/5688080/31378270-32715e8e-adc4-11e7-9cac-607fd6efc51d.png)

There is no marker labels, only tooltips. I think it is OK, we need own Icon implementation any way for our SVG images, can add label to it.

**Comment by tananaev** [comment]
I'm thinking of splitting out images into background and icon. This way people would be able to easily use custom icons.

**Comment by Abyss777** [comment]
Agree, It is reasonable and we will not need define IDs for every icons.
Also I'm thinking, there are some FR to show motion state and ignition on the Icon. May be it makes sens to envisage this functionality. Like small overlay icon of key (ignition) and some arrow or exclamation mark (to show motion). What do you think?

**Comment by tananaev** [comment]
Sure, we need to make it flexible. Probably doesn't make sense to overcomplicate everything from the beginning, but we need to keep it in mind.

**Comment by Abyss777** [comment]
OK. I'll start preparing some draft for device icon class for leaflet.

**Comment by Abyss777** [comment]
Actually it is easy to use font awesome directly, like this https://github.com/danwild/leaflet-fa-markers
Can we use `icon8` in the same way?

**Comment by tananaev** [comment]
I'm not sure. It might be too big to include as a font.

**Comment by Abyss777** [comment]
Implemented a draft
https://drive.google.com/open?id=0Bxu1vjhFvq74MDdKQnJETHJYN1U

Works smoothly and very fast, nothing blinks.
Icons is two SVG: internal icon loaded from files and background inlined. Rotating happens on SVG, without replacing whole icon. Zooming happens on `div` container. Caption placed as `span`

Ready to show the code somehow.

**Comment by tananaev** [comment]
Looks nice.

**Comment by Abyss777** [comment]
Implemented rather functional simple application with leaflet and pure js

https://github.com/Abyss777/traccar-web/tree/traccar-web-new/web/new

![image](https://user-images.githubusercontent.com/5688080/31539676-c770e328-b022-11e7-988d-c3f855efcdb8.png)

It supports device images, openstreetmap, carto and custom maps (for others we probably should use plugins), also implemented following selected device.
Tested in Chrome/Chromium, FF and IE 11 (Edge)

**Comment by tananaev** [comment]
Perfect as a reference, but we need to layout some architecture for the whole app to make it modular and flexible.

**Comment by Abyss777** [comment]
Totally agree, I've just touched leaflet to understand how easy/hard to work with it. Actually I liked it. (But I hate zoo of browsers)

Yes we need to choose a way to work with modules and dependencies. May be it is reasonable to use the same as in libraries we going to use. Or others...

http://requirejs.org/docs/whyamd.html
https://github.com/umdjs/umd

**Comment by tananaev** [comment]
I'm on holiday till 14th, but after that I should have more time to work on the new web app. We definitely need to decide how we are going to deal with modules and libraries.

**Comment by tsmgeek** [comment]
Have you seen the potential issues with React due to the Facebook licence. It could cause any company issues. WP have decided to pull it all out and implement vue.js instead.

Should be some though to how big some of the libraries are, extjs is really heavy which is slow to init on mobile devices.

Its all well and nice to try having one responsive interface for desktop/mobile but ultimately what works well on desktop does not work so well on mobile. It may be an idea to create a full/light interface that reuses much of the same core code.

**Comment by tananaev** [comment]
Have you read the thread before commenting? We are not planning to use React.

Obviously it's good to have separate desktop and mobile websites. Even better would be to have native mobile apps, but we just don't have resources to support it. The only feasible option I see at the moment is responsive app. Feedback on Traccar Manager clearly shows that slower app with full functionality is better than fast native app with less functionality.

**Comment by tsmgeek** [comment]
I was only putting another reason to why react could/would be an issue if there was any backtrack later.

Im not talking about a whole other app, just a "lite" interface, it may be that the new design will be more responsive than the current extjs implementation of panels which does not work well on mobiles.

**Comment by tananaev** [comment]
That's one of the main goals. Make fast and responsive app that works fine on both desktop and mobile.

**Comment by kirantpatil** [comment]
My vote goes for Vue.js.

Please refer, https://vuejs.org/v2/guide/comparison.html

Learning Curve

To get started with Vue, all you need is familiarity with HTML and ES5 JavaScript (i.e. plain JavaScript). With these basic skills, you can start building non-trivial applications within less than a day of reading the guide.

**Comment by tananaev** [comment]
As far as I understand Vue is very similar to React, so it doesn't really fit very well here.

**Comment by kirantpatil** [comment]
For front end app development ( modularize) , IMHO it is better to use [webpack](https://webpack.js.org/) and [Yarn](https://yarnpkg.com/en/) [Yarn](https://code.facebook.com/posts/1840075619545360) for dependency management.

This will help, if tomorrow the project wants to switch to other framework, webpack will help for smooth  transition.

Rails framework has switched to webpack so that they can support the upcoming frontend frameworks which is taken care by webpack.

**Comment by kirantpatil** [comment]
@tananaev, any updates on development stack, direction and tools to build new Frontend Application ?

**Comment by tananaev** [comment]
No updates yet.

**Comment by lambert11** [comment]
Little inspiration: [http://geliospro.com](http://geliospro.com)

**Comment by allenhwkim** [comment]
If your intention is to find framework-independent UI components,  I have made custom elements without framework dependancy. You can make use of it by adding each or all to your website, https://github.com/allenhwkim/custom-element.

It is intended to work modern browsers FF, Chrome, Safari, and IE, but tested without IE.

It's still under development,  but any feedback would be appreciated.

**Comment by tananaev** [comment]
The plan is to use some mature project with big community behind.

**Comment by MAJLGC** [comment]
What about Angular 5 and Material Design?

**Comment by tananaev** [comment]
What I don't like about Angular is that it's too opinionated. We are going to be tied to another framework forever again.

**Comment by MAJLGC** [comment]
At least a Typescript enabled framework maybe considered

2017-11-29 10:15 GMT+01:00 Anton Tananaev <notifications@github.com>:

**Comment by kirantpatil** [comment]
How about using stimulus js.

https://github.com/stimulusjs/stimulus

A modest JavaScript framework for the HTML you already have

Stimulus is a JavaScript framework with modest ambitions. It doesn't seek to take over your entire front-end—in fact, it's not concerned with rendering HTML at all. Instead, it's designed to augment your HTML with just enough behavior to make it shine. Stimulus pairs beautifully with Turbolinks to provide a complete solution for fast, compelling applications with a minimal amount of effort.

**Comment by tananaev** [comment]
This description is completely useless. It doesn't explain anything.

**Comment by ovnicraft** [comment]
Any decision taken about this ?

**Comment by BinoJozef** [comment]
Hi .

I just saw your conversation. i'm ready to help if you choice angular 5 (progressive web apps).

**Comment by sebdigo** [comment]
Hi.

Any update on the new framework ?

**Comment by tananaev** [comment]
I'm investigating Materialize CSS framework for the UI, but it's not final decision yet.

**Comment by Malouda** [comment]
Any update!?

**Comment by mhuzqu** [comment]
Hii all,
I wonder i can help in this project.
Currently i try to understand the source traccar web.

I know one project fork from traccar web called traccar web ui mod http://traccar.litvak.su/index.html written by Vitaly Litvak, he project looks well.

Consider use Framework7.io, have greate native mobile material apps.

**Comment by leifnel** [comment]
It seems no progress on litvak in two years.

**Comment by mail2bishnoi** [comment]
Any Update on Front End Framework ?

**Comment by LeonAlvarez** [comment]
I would go with react, and yes u can use non react libraries with react just building a wrapper, when app get bigger and complex redux is really good, help handling inmutability.

**Comment by kirantpatil** [comment]
I agree with @tananaev, earlier we faced same issue in one of our internal project. We jump started with ember and failed to get good map library and finally reverted back to pure javascript, jquery and leaflet and our productivity skyrocketed.

**Comment by ovnicraft** [comment]
after all comments the path is clear: pure javascript is the solution, the question here is: there is a roadmap for this ?  and community can decide to help ?

Asking about updates at this point dont help so much, i am not frontend developer but consider experience people can start with a PR :-) 

regards !

**Comment by mail2bishnoi** [comment]
I am ready to work. Tananaev Which framework you are using ?

**Comment by tananaev** [comment]
@mail2bishnoi, if you asking this question it means that you haven't read this thread. Please read it first before asking questions.

**Comment by mail2bishnoi** [comment]
I have read it but I am not sure which one you have finalized-
pure java script + material-ui + Leaflet;

**Comment by ovnicraft** [comment]
@tananaev i hope you can find time and push a PR :-)

**Comment by timaschew** [comment]
Just my 2 cents:
I've worked with React (+Redux)  with Vue.js and now I started a new project in [markojs](https://markojs.com). IMHO it's much simpler and easier to use than React and Vue.js.
I really like that you can have still full control via low level API.

**Comment by r270ba** [comment]
@tananaev I am getting ready to put a full time developer on building a new web interface for Traccar into our software.   I would love to help contribute back to the project.  I have read all of the posts and I am also unclear as to the decided direction.  I see your last mention of Materialize CSS framework but that nothing was finalized.  Let me know how our team can contribute.

**Comment by tananaev** [comment]
@r270ba please send me an email to discuss it.

**Comment by mail2bishnoi** [comment]
@tananaev I am also ready to work for working on new web interface. Also sent you am email, please share  the details.

**Comment by tananaev** [comment]
I have decided to give React another chance. You can find some initial experiments here:

https://github.com/traccar/traccar-web/tree/modern/modern

Feedback and contributions are welcome.

**Comment by oliv3** [comment]
I'm trying to keep from being overly passionate or enthusiastic (after all, this is just an announcement), but a chance to integrate Cesium ? That would be pretty darn cool.

* [Cesium](https://cesium.com/index.html)
* [Blog article on Cesium+React](https://cesium.com/blog/2018/03/05/integrating-cesium-and-react/)
* [cesium-react](https://github.com/rot1024/cesium-react)

I'll give a look to the experiments for sure.

Cheers,

**Comment by tananaev** [comment]
It seems to be proprietary software and it's not even free.

**Comment by oliv3** [comment]
https://github.com/AnalyticalGraphicsInc/cesium/blob/1.48/LICENSE.md

**Comment by tananaev** [comment]
I don't understand what is this then:

https://cesium.com/pricing/

Anyway, it doesn't seem like React wrapper is actively supported.

**Comment by tananaev** [comment]
I have added a simple login flow and also a short readme for those who interested in trying it our or helping with development:

https://github.com/traccar/traccar-web/tree/modern/modern

**Comment by oliv3** [comment]
Can we expect server API commands to be synchronized ?

**Comment by oliv3** [comment]
It's not about API changes - the API is fine as it is.
It's about the web interface not being synchronized with actions done through the API, other than location updates or events.
There are many users complaining that some actions done via the API are simply not reflected in the interface, and that you have to **reload the page** to see them taking effect.
So this looks like the right place to ask.
If keeping the interface not synchronized with everything else doesn't seem a high priority to you, then well, I'm not interested in it.

**Comment by tananaev** [comment]
Not, it's not the right place to ask. If there is some specific issue, we can discuss it in a separate thread. This is for more high level architecture discussion.

**Comment by oliv3** [comment]
There's already https://github.com/traccar/traccar-web/issues/683, then.

**Comment by oliv3** [comment]
There, https://github.com/traccar/traccar/issues/4044.

**Comment by tananaev** [comment]
This is what the web interface looks like at this stage:

![screen shot 2018-09-08 at 11 39 16 am](https://user-images.githubusercontent.com/1161863/45247490-27268880-b35c-11e8-95d6-d33f772b32af.png)

We are very far from parity with old web app, but I think soon it will be possible to use new web app for simple viewing.

Feedback and contributions are welcome.

**Comment by jike212** [comment]
Another alternative to Material for something like this might be Blueprint JS from Palantir (has React components): https://blueprintjs.com/

It's known for being good with data-dense applications but I'm not sure about it's responsiveness.

**Comment by tananaev** [comment]
Have you actually looked at the code? The app is trivial at the moment. It's just pure React. There is plenty of documentation available on React.

**Comment by tananaev** [comment]
Also, there is documentation with all the references that you need:

https://github.com/traccar/traccar-web/tree/modern/modern

**Comment by Malouda** [comment]
why dont you create a new repository for this new ui

**Comment by ovnicraft** [comment]
@Malouda new repo will confuse, i guess @tananaev will update web repo when cosider ready.

So i open a PR https://github.com/traccar/traccar-web/pull/719 updating readme with this changes to inform community.

**Comment by mohammedi-haroune** [comment]
I tested it, it looks very nice!
What's the state of the development of this is new web app version, is there any road map for this?

**Comment by tananaev** [comment]
I don't have much time to work on it at the moment, so the progress is very slow.

**Comment by mohammedi-haroune** [comment]
@tsmgeek What if I want to work on it given that I haven't any experience with reactJs, I have some experience with Vuejs thought ! how much time would it take and how can I proceed the development.

**Comment by ovnicraft** [comment]
@tananaev what about run a crowdfunding ?

**Comment by tananaev** [comment]
It would definitely help with motivation, but the biggest problem at the moment is lack of time to work on it.

**Comment by mohammedi-haroune** [comment]
@tananaev What if I want to work on it given that I haven't any experience with reactJs, I have some experience with Vuejs thought ! how much time would it take and how can I proceed the development.

**Comment by tananaev** [comment]
I don't have any previous experience with ReactJS either. The app is really trivial at this point, so it should be easy to start.

**Comment by mohammedi-haroune** [comment]
OK! I'll give it a try. Thank you

**Comment by mohammedi-haroune** [comment]
@ovnicraft I was looking on the code and I think the application will mainly use the [traccar-api](https://www.traccar.org/traccar-api/) to fetch and post data, I'm I wrong ?

**Comment by ovnicraft** [comment]
AFAIK yes !

**Comment by tananaev** [comment]
Yes, you should use same API. We can modify or add API calls if really needed, but preference is to keep API as it is.

**Comment by Urjashee** [comment]
I propose that we should write the modern UI with Vue.js . I infact have some code in development . With vuetify as the ui framework I think its the best way to go .

**Comment by mohammedi-haroune** [comment]
@Urjashee Why do you think this is the best way to go?
I'm personally afraid that VueJs is not as mature as React and we can't find things like vue-leapfeat for the map libraray ... etc

**Comment by ovnicraft** [comment]
@mohammedi-haroune Vue in totally accepted as mature an stable, that is not doubt. 
So i propose this:
Write a PoC and evaluate it. 

We can discuss many time and do nothing, lets do it my friends !

**Comment by Malouda** [comment]
@ovnicraft Vuejs for this project is not a very good idea,since vue is on a route to version 3 which will break backward compatibiliby...react is the best choise

**Comment by parthibd** [comment]
@Malouda @mohammedi-haroune Yes, Vue is on its way to version 3. But backward compatibility will be there for majority of the stuff . 

The reason to choose Vue is that it is very mature and stable . Much like react .

**Comment by parthibd** [comment]
And Vue also has a very mature UI framework , Vuetify . I see why not we shouldn't use Vue .

**Comment by mniyaz** [comment]
Dear @tananaev, is there a way to deploy this?, because i coded some UI and want to see it run with my existing Traccar server.

**Comment by tananaev** [comment]
Of course there is. You would need to set up nodejs and configure proxy to route all requests correctly.

**Comment by mniyaz** [comment]
Thank you. Can you help me with file to change proxy. I tried changing websocket and it worked in local, but not when deployed.

**Comment by parthibd** [comment]
Check your firewall configuration.

**Comment by mniyaz** [comment]
It was not firewall issue, deployment was successful. Device list is working, but positions doesn't render consistently, sometimes only online vehicles displayed on map.

**Comment by ilstarno** [comment]
is there any progress on this porting?

**Comment by tgroshon** [comment]
@tananaev I am interested in helping out with the client rewrite with React + Leaflet.  I looked at the prototype currently in the repo and I've done some similar things.  

Do you have a roadmap of the features and progress you need?

**Comment by oliv3** [comment]
https://github.com/traccar/traccar-web/graphs/code-frequency

**Comment by tananaev** [comment]
@tgroshon the plan is implement all or most of the features available in the original web app. There is no specific roadmap. Any help is welcome.

**Comment by oliv3** [comment]
@tananaev Any good reason why my comment was censored ?
Linking to the project's activity is a way to show it needs attention.

**Comment by GoZippy** [comment]
Is this /modern still being considered the future implementation? Where are you at with it now - read through all the above and not sure what needs to be done or how I can help now.

**Comment by tananaev** [comment]
@GoZippy, yes, it is still considered the future implementation. We need to implement all the same features that we have in the current web app.

**Comment by spacecabbie** [comment]
> @GoZippy, yes, it is still considered the future implementation. We need to implement all the same features that we have in the current web app.

Like to test and see if i can help out witch repository is the current one in ? modern tree last update was 8 months ago ?

**Comment by tananaev** [comment]
@spacecabbie it's the correct branch. Haven't pushed anything for a long time, but I have been doing some local experimentation.

**Comment by oliv3** [comment]
What prevents from pushing commits in a "WIP" branch ? As @spacecabbie said, 8 months without any activity may make people think this is an abandoned project.

**Comment by tananaev** [comment]
Because there is nothing to commit, obviously.

**Comment by oliv3** [comment]
Do you mean your local experimentations are not worth sharing ?

**Comment by r270ba** [comment]
Oliv3 tone it down a bit. Current product is awesome. The undertaking to rebuild the interface from the ground up is huge. I know because I have 3 full time developers building mapping and tracking applications. 

If you're not careful it really will get abandoned because developer will get tired of dealing with complaints. 

Obviously if he had something worth sharing he would push to the repo.

**Comment by xpressconsulting** [comment]
Have you considered using GraphQL with this? You could then have one endpoint and it would be super fast. I would also like to contribute as well. I have been using traccar since the 2.0 release. Anton has always been a huge help when I was just getting started years ago. And Vitaly was also super helpful with the UI mod and docs. Glad that the team came together.

**Comment by tgroshon** [comment]
GraphQL is cool, but I don’t think it would necessarily improve things that much. As it stands, the Websocket API is accessible and used by the web and mobile clients without real issue. The problem now is that one client, web, needs to be reimplemented because it’s underlying framework is being discontinued.

**Comment by xpressconsulting** [comment]
I don't disagree with that. Was just asking because if you are going through the effort of doing the front end, may as well do that also. Just my thought...

**Comment by ovnicraft** [comment]
Its possible define a team to do it, we have so many time discussing it.

**Comment by josuelmm** [comment]
Good day, I would like to support in the development of the project in its version of traccar web react. How can I help you?

**Comment by tananaev** [comment]
Start implementing features and send pull requests.

**Comment by josuelmm** [comment]
They have already started with something or everything from scratch. What scripts, plugins, complements, have weighed, as you already analyzed more than me, I would like to know that.??
josuelmm92@gmail.com

**Comment by tgroshon** [comment]
I started doing some simple work on the React + Redux + Material UI rewrite that @tananaev started.  To be a _loose guideline_, I made a public trello board with some obvious user stories that need tackled: 

https://trello.com/b/kPlpdZD4/traccar-web-modern

The list of tasks is not comprehensive nor detailed, but it is something.  Anyone can view the board and cards, but editing requires you to be added to the team.  I am willing to add anyone to the Trello team who wants to be actively involved in the development.  You can contact me at [tom@groshong.software](mailto:tom@groshong.software)

**Comment by tgroshon** [comment]
FYI @tananaev is also a Trello team and board admin with ability to add anyone and change anything.

**Comment by oliv3** [comment]
Oh. Like rewriting history and/or deleting posts ?

**Comment by pycc34** [comment]
how to do geofence with react app??

**Comment by ilstarno** [comment]
https://github.com/dev-labs-bg/geofencing/blob/master/README.md

it is already an port of it on react it seems.

not tested it but i guess it should work

**Comment by jinzo** [comment]
So a month ago I finally had some time to play with React and React Native for a new/modern web interface. So if I understand it right noone else is working on it right now? 
I usually avoid doing many big changes in one go and prefer incremental changes - but in this instance I think it would be beneficial if the API and Socket interface (I'll just refer to both of them as API from now on) of Traccar change with the new paradigm. I think there are some pain points in the API that would make developing and maintaining a React "app" easier. As already mentioned in traccar/traccar#4336 a newer way of logging in is probably a good idea, especially with the 2FA/U2F in mind. But that's not that critical. A bit of a bigger pain point was the fact that the Socket api sends an array of all the _devices_ on every new position reported, but doesen't send a new array when for example, the device name changes. So I think a better design would be that the Socket API emits a message for the specific device when any (at least user facing) field changes (so no array of all devices for one device change and a message for every change not only last location update). This would make state management in React app a lot easier, as we would only have one source of changes for the devices/positions. I don't know if @tananaev is ok with substantial changes to the API?
Note - I did most of my work on 4.3 but I don't think anything regarding the API changed in the newer versions. 

Regarding the app/functionality. I think I found quite a nice solution for the React Native/Mobile app layout (at least the most commonly used device list/map/reports). I'll whip out some wireframes over the weekend and post them here for feedback. The progress will probably be slow, but I think I'll finally be able to do at least a consistent amount of work towards a new interface the following weeks/months.

**Comment by tananaev** [comment]
I don't think what you are proposing is relevant to the new web app.

**Comment by jinzo** [comment]
Yeah it's not entirely connected with the new web app, but it's something I noticed when doing a React Native prototype. I'm far from an experienced React developer, but I think that changes like that would make it a lot easier to develop a React app. But I guess they're not required, at least for now. I'll post the wireframes for the mobile app and so we can discuss the general layout and I can continue with the prototype in the right direction.

**Comment by GoZippy** [comment]
@jinzo make any progress? I want to help on this badly but am so taken of time with work and family it's driving me nuts. Know anyone I can offer up some cash to for help on this project and maybe helping me build what I am thinking this would be a great base for? I checked out the Trello board - not seeing anything in there - but I don't use Trello - ever - so willing to learn... looked over at the github rep for modern and do not see any activity there either.. Guys/gals - what can I do to help or offer up as a rewards base to attract some talent to this project? I dont have much time to work on this directly - I will help as I can though. I figure my best offer is to identify approved direction plans and milestones and offer up a bounty to anyone willing to put in the time to make the pull requests and commits... any thoughts?

**Comment by tananaev** [comment]
The plan is simple - implement same functionality that we have in the original app.

**Comment by nilocoelhojunior** [comment]
Hi, how can i help?

**Comment by GoZippy** [comment]
Any decisions made on where to go with the web interface yet? I have not seen any trello updates or notices here for a while... Let me know what to work on and I can definitely put some time in for this project.

**Comment by dt-ap** [comment]
I built an little example using react and typescript:
https://github.com/dt-ap/traccar-react

Currently, the app could do a simple login, getting devices, receive position from websocket, and generate trips' reports.

I would love to help in my spare time for the modernisation of traccar web.

**Comment by tananaev** [comment]
Here's a build of the new web app: [modern.zip](https://github.com/traccar/traccar-web/files/4367458/modern.zip). Replace "web" folder with the content of the archive to try it on your server. The functionality is very basic at the moment, but I think it's a good starting point.

The "modern" branch is now merged into master and I plan to include the new web interface as an option in the next official release.

**Comment by spacecabbie** [comment]
> Here's a build of the new web app: [modern.zip](https://github.com/traccar/traccar-web/files/4367458/modern.zip). Replace "web" folder with the content of the archive to try it on your server. The functionality is very basic at the moment, but I think it's a good starting point.
> 
> The "modern" branch is now merged into master and I plan to include the new web interface as an option in the next official release.

Anything you want tested or done ? Installed and worked straight out of ta box. Looks very nice kudo's for all that made it possible.

**Comment by tananaev** [comment]
Testing is good. If there are any issues, please report. If anyone has time, it would be nice to start adding more functionality to eventually replace the old web app with this one.

**Comment by spacecabbie** [comment]
> Testing is good. If there are any issues, please report. If anyone has time, it would be nice to start adding more functionality to eventually replace the old web app with this one.

Sure atm it runs fine but there is also very little to break :P. Love to give it a go to help but i have no access to a java compiler (I presume its still needed.) I could play around with the source and commit uncompiled i suppose the source is and will remain at: https://github.com/dt-ap/traccar-react ?

**Comment by tananaev** [comment]
No need for Java compiler. It's a standard React application. The source code is here in this repo.

https://github.com/traccar/traccar-web

**Comment by spacecabbie** [comment]
> No need for Java compiler. It's a standard React application. The source code is here in this repo.
> 
> https://github.com/traccar/traccar-web

Yep was looking in the wrong place cloned it and am running it this seems doable for a novice like me. I could make some base layouts menu etc i guess. I'll have to find a visual tool for it.

**Comment by Alexander090191** [comment]
Hello, I am a full stack software developer, I am interested in this project, I think it is paused and I would like to contribute to the development of the new application.

**Comment by tananaev** [comment]
It's not really paused. Feel free to contribute.

**Comment by tananaev** [comment]
If anyone is interested in contributing to the new Traccar web app, there's now an opportunity to [get a free t-shirt](https://hacktoberfest.digitalocean.com/) for it. Feel free to contact me if you need some pointers on what things you can work ok. It can be as small or big as you want.

**Comment by dt-ap** [comment]
Sounds good, really interesting. It would be nice if you have some list of small or big features/pointers/bugs that we could maybe worked on

**Comment by tananaev** [comment]
Smaller more straightforward work:

- Migrate drivers
- Migrate computed attributes
- Migrate saved commands
- Migrate maintenance
- Implement report exports

For the above features you can just follow existing patterns.

A bit more complicated:

- Migrate calendars
- Implement other reports (require refactoring to avoid code duplication)
- Migrate geofences (this is probably hardest)
- Add support for category icons

**Comment by dt-ap** [comment]
Is there a running traccar server with some data for staging/testing? Because I don't have a running traccar server, but would love to work on the web interface.

**Comment by tananaev** [comment]
You can use demo servers if you want. You don't need any data to implement most of the features that I mentioned.

**Comment by tananaev** [comment]
Modern app is officially out.
